Distributed Tracing

TsuyoshiUshio opened the pull request:
Hi @cgillum 

I'm creating this pull request to have this branch to hands off to you. 
I open this PR for visibility. 

# What I've done: 
* Correlation design for Orchestration and implementation
* Working Sample

# TODO 
* Error Tracking for throwing exception case. 
* Continuous as New design
* Durable Functions Extension contribution

Now one of the test case will fail for not implementing it for multiple orchestration retry case. However, I open this for share.

TsuyoshiUshio commented:
![correlated](https://user-images.githubusercontent.com/1390976/54005619-f955ad00-410e-11e9-89c9-4258514c76aa.JPG)

# Done
* Correlation with Orchestration
* Correlation with Multiple Orchestration
* Correlation with Exception
* Correlation with Retry (Activity)
* Correlation with Retry (Orchestrator)
* Multi Layered Orchestration
* Multi Layered Orchestration with Retry
* Fan-Out Fan-In 

# TODO
* Continue As New Testing
* Automatic tracked Dependency with correlation
* Correlation with Http Correlation Protocol ( Currently only for W3C TraceContext)
* Activity Restoration (After .NET implement it)

cgillum commented on src/DurableTask.Core/TaskOrchestration.cs:
```diff
@@ -132,7 +132,7 @@ public override string GetStatus()
         /// <param name="input">The typed input</param>
         /// <returns>The typed output</returns>
-        public abstract Task<TResult> RunTask(OrchestrationContext context, TInput input);
+        public abstract Task<TResult> RunTaskAsync(OrchestrationContext context, TInput input);
```
This is a breaking change, which cannot be allowed.

TsuyoshiUshio replied:
Hi @cgillum 

Sorry, I didn't notice that. However, I might mistake to refactor the name.  I recover it.

TsuyoshiUshio commented:
Hi @cgillum The CI fails, it looks the script can't refer the $ServiceBusConnectionString . Hmm I have no idea. Any thoughts? If not, I'll ask the Azure DevOps guys.

cgillum commented:
@TsuyoshiUshio I'm not sure. The CI passed earlier today for a different PR. I have not made any changes to the pipeline.

TsuyoshiUshio commented:
@cgillum I'm asking the Azure DevOps guys.

simonporter commented on Correlation.Sample.Tests/StringExtensionsTest.cs:
```diff
@@ -0,0 +1,38 @@
+using Microsoft.VisualStudio.TestTools.UnitTesting;
```
>using [](start = 0, length = 5)

File header + move using into namespace

TsuyoshiUshio replied:
Done.

cgillum replied:
Looks like this still needs to be done?

simonporter commented on src/DurableTask.Core/TaskActivityWorkItem.cs:
```diff
@@ -11,6 +11,8 @@
 //  limitations under the License.
 //  ----------------------------------------------------------------------------------
 
+using System.Diagnostics;
```
>using [](start = 0, length = 5)

move into namespace

TsuyoshiUshio replied:
Done

simonporter commented on src/DurableTask.Core/CorrelationTraceContext.cs:
```diff
@@ -0,0 +1,107 @@
+    {
+
+#if NETSTANDARD2_0
+        private static AsyncLocal<TraceContext> current = new AsyncLocal<TraceContext>();
```
>private [](start = 8, length = 7)

don't use private modifier per rest of repo convention

TsuyoshiUshio replied:
Hi @simonporter , 
Before fix this, Could you clarify the convention?  Should I use internal or default for the modifier in this case?  Why do you forbid private modifier? I'd like to know the reason for the future contribution of this repo.  Also want to properly fix this.

TsuyoshiUshio commented:
One note. I'm adding `Microsoft.ApplicationInsights`  namespace on the DurableTask.Core. It is just temporary. It is used for ActivityExtension which enable W3C TraceContext. Application Insights team move the extension to the System.Diagnostics namespace. The PR is already merged 15 days before. It will be included the next version of the .Net Core implementation. https://github.com/dotnet/corefx/pull/33207

TsuyoshiUshio commented:
Hi @simonporter , 

Sounds good for the convention? I'm working with @cgillum for enabling Distributed Tracing for Durable Functions. However, we try to make it available for other teams that use DurableTask. This PR is not definitive version. If you don't like my implementation, you can dispose it. However, the architecture and what I experienced might be helpful at least for the future implementor. Could you give me 30 min to explain this PR in person? I'll come. (Hopefully with @cgillum . ) This PR might be too big for review in here.

TsuyoshiUshio commented:
The CI fails are the CI issue. Not code. the lack of permission for public PR. 
I'm investigating it with my colleague.

ShreyasRmsft commented:
@TsuyoshiUshio thank you for bringing this to our notice. We (VsTest team) are working on getting distributed test runs to work with public projects. Will keep you updated.

TsuyoshiUshio commented:
Hi @ShreyasRmsft , Thank you for telling me! Sure. Sounds exciting. until then we just add PR contributors some kind of group.

TsuyoshiUshio commented:
@ShreyasRmsft Could you tell me the best place to file this issue? I spent several days for this issue. I don't want people have the same experience as me. Also, if it is OK.  could you tell me when it is going to be published? I also MSFT guy. However, I can't identify you from your GitHub account. :)

ShreyasRmsft commented:
@TsuyoshiUshio you can raise the issue on https://github.com/Microsoft/azure-pipelines-tasks/issues

TsuyoshiUshio commented:
Thank you, @ShreyasRmsft , I've done that!

TsuyoshiUshio commented:
Hi @simonporter 

This is the sample of the raw message with Two Layered OrchestratorTraceContext.  Please see the `TraceContextStore` part. I change the serialization option only for this part for supporting reference. I don't want to change the whole serialization option. Only for `TraceContextStore` part. However, if you have any idea to reduce the message or put out of queue, please let me know.  around 1500 bytes. 

```
{"$type":"DurableTask.AzureStorage.MessageData","ActivityId":"76185f4d-bb4a-4ed4-857e-3e9eb1e42690","TaskMessage":{"$type":"DurableTask.Core.TaskMessage","Event":{"$type":"DurableTask.Core.History.TaskCompletedEvent","EventType":5,"TaskScheduledId":0,"Result":"\"Hello world with retry\"","EventId":-1,"IsPlayed":false,"Timestamp":"2019-03-21T17:48:43.5263421Z"},"SequenceNumber":0,"OrchestrationInstance":{"$type":"DurableTask.Core.OrchestrationInstance","InstanceId":"67b1704f68b049ae9d415d58aa0169c9:2","ExecutionId":"163b7b9a6efe40d586a68f844f7461b4"}},"CompressedBlobName":null,"SequenceNumber":9,"Episode":1,"TraceContextStore":{"$type":"DurableTask.Core.TraceContextStore","TraceContextJson":"{\"$id\":\"1\",\"$type\":\"DurableTask.Core.TraceContext, DurableTask.Core\",\"StartTime\":\"2019-03-21T17:48:43.526875+00:00\",\"ParentId\":\"|568a23ca-42c37626d2275a9f.568a23cf_568a23fc_568a2451_568a2456_568a2458_\",\"Traceparent\":\"00-c91fda74588ce04693a54a2ce6919c56-bab42c068a951740-02\",\"Tracestate\":null,\"ParentSpanId\":\"0463ebaf65d93843\",\"OrchestrationTraceContexts\":[{\"$id\":\"2\",\"$type\":\"DurableTask.Core.TraceContext, DurableTask.Core\",\"StartTime\":\"2019-03-21T17:47:26.5219035+00:00\",\"ParentId\":\"|568a23ca-42c37626d2275a9f.568a23cf_568a23fc_\",\"Traceparent\":\"00-c91fda74588ce04693a54a2ce6919c56-2158e5eba664fb4a-02\",\"Tracestate\":null,\"ParentSpanId\":\"cf0bcc1ba2f5f34f\",\"OrchestrationTraceContexts\":[{\"$ref\":\"2\"}],\"RootId\":\"00-c91fda74588ce04693a54a2ce6919c56-2158e5eba664fb4a-02\"},{\"$id\":\"3\",\"$type\":\"DurableTask.Core.TraceContext, DurableTask.Core\",\"StartTime\":\"2019-03-21T17:48:43.4696347+00:00\",\"ParentId\":\"|568a23ca-42c37626d2275a9f.568a23cf_568a23fc_568a2451_\",\"Traceparent\":\"00-c91fda74588ce04693a54a2ce6919c56-d84f945c3a626642-02\",\"Tracestate\":null,\"ParentSpanId\":\"2158e5eba664fb4a\",\"OrchestrationTraceContexts\":[{\"$ref\":\"2\"},{\"$ref\":\"3\"}],\"RootId\":\"00-c91fda74588ce04693a54a2ce6919c56-d84f945c3a626642-02\"}],\"RootId\":\"00-c91fda74588ce04693a54a2ce6919c56-bab42c068a951740-02\"}"}}
```

TsuyoshiUshio commented:
Hi @simonporter , @cgillum , 

I implement the test for ContinueAsNew.  The current behavior is that Orchestration which is created by ContinueAsNew is considered as the same Orchestration. It is because of the `isReplay` is true  after the ContinueAsNew.  If you don't like this behavior,  we might change the behavior that if an orchestration is continueAsNew, then start a new orchestration. However, current behavior looks good as well. 

![image](https://user-images.githubusercontent.com/1390976/55194298-78666000-5166-11e9-9bcd-46842cfc8198.png)

adarsh1 commented on src/DurableTask.Core/TaskOrchestrationContext.cs:
```diff
@@ -341,6 +341,11 @@ public void HandleTaskFailedEvent(TaskFailedEvent failedEvent)
                     failedEvent.Reason, cause);
 
+                // correlation
+                var traceContext = CorrelationTraceContext.Current;
```
are traceContext and activity used?

TsuyoshiUshio replied:
Removed

adarsh1 commented on src/DurableTask.Core/TaskHubClient.cs:
```diff
@@ -487,9 +503,19 @@ public Task<OrchestrationInstance> CreateOrchestrationInstanceAsync(string name,
                 });
             }
 
+            // TODO Good place to add the dependency telemetry and track Activity.
```
IS the todo still needed?

TsuyoshiUshio replied:
removed

adarsh1 commented on src/DurableTask.Core/CorrelationTraceClient.cs:
```diff
@@ -0,0 +1,125 @@
+    /// </summary>
+    public class CorrelationTraceClient
+    {
+        const string DiagnosticSourceName = "DurableTask.AzureStorage";
```
This shouldn't be hardcoded to the Azure Storage Provider in Core.

cgillum replied:
This still needs to be fixed.

cgillum replied:
The current project is DurableTask.Core. It should have no knowledge of DurableTask.AzureStorage. Either you need to change this to `"DurableTask.Core"` or you need to make it dynamic. The correct approach depends on your design.

adarsh1 commented on src/DurableTask.Core/TaskHubClient.cs:
```diff
@@ -487,9 +503,19 @@ public Task<OrchestrationInstance> CreateOrchestrationInstanceAsync(string name,
             await ServiceClient.SendTaskOrchestrationMessageBatchAsync(taskMessages.Where(t => (t.Event is EventRaisedEvent)).ToArray());
+
+            // Correlation
+            CorrelationTraceClient.TrackRequestTelemetry(requestActivity);
```
Should this be before we actually send the Create Task Orchestration message? considering there could be an exception thrown from CreateTaskORchestrationAsync.

TsuyoshiUshio replied:
Generally speaking, Request Telemetry include StartTime and Duration. Ideally, Duration should be between starttime and end time of the method. So I include sending OrchestrationMessage. 

However, I follow your advice. You mean ServiceClient.CreateTaskOrchestrationAsync could throw an Exception on the normal case, right? Done.

adarsh1 commented on src/DurableTask.Core/StackExtensions.cs:
```diff
@@ -0,0 +1,26 @@
+        /// <returns></returns>
+        public static Stack<T> Clone<T>(this Stack<T> original)
+        {
+            var arr = new T[original.Count];
```
Would something like new Stack(original.Reverse()) here make sense? to avoid allocating a new Array.

TsuyoshiUshio replied:
Done. Thank you. It is much better way.

adarsh1 commented on src/DurableTask.Core/ActivityExtensions.cs:
```diff
@@ -0,0 +1,111 @@
+        /// <param name="activity"></param>
+        public static void SetActivityCurrent(this Activity activity)
+        {
+            var property = typeof(Activity).GetProperty("Current", BindingFlags.Public | BindingFlags.Static);
```
Why this instead of simply activity.Start()?

TsuyoshiUshio replied:
I move this part to the TraceContextBase. It is required. Start() method include update startTimeUtc as UtcNow.  This usecase is, when we want to put Activity.Current, it is already started on the different call graph.

adarsh1 commented on src/DurableTask.AzureStorage/AzureStorageOrchestrationService.cs:
```diff
@@ -656,15 +659,43 @@ static TaskHubInfo GetTaskHubInfo(string taskHub, int partitionCount)
                         return null;
                     }
 
+                    // Correlation TODO: Need to understand the actual behavior of the multiple message.
```
Is this TODO still needed?

TsuyoshiUshio replied:
Removed

adarsh1 commented on src/DurableTask.AzureStorage/AzureStorageOrchestrationService.cs:
```diff
@@ -656,15 +659,43 @@ static TaskHubInfo GetTaskHubInfo(string taskHub, int partitionCount)
+                    }
+                    else
+                    {
+                        // TODO this peek sometimes fail. Find out the root cause.
```
Is this TODO still needed?

TsuyoshiUshio replied:
Removed

adarsh1 commented on src/DurableTask.AzureStorage/AzureStorageOrchestrationService.cs:
```diff
@@ -833,17 +864,65 @@ async Task AbandonAndReleaseSessionAsync(OrchestrationSession session)
+            TraceContext currentTraceContextOnComplete = null;
+
+            // TODO Terminated and Canceled might be the same. 
+            if ((orchestrationState.OrchestrationStatus == OrchestrationStatus.Completed) ||
```
should this be changed to OrchestrationStatus!=Pending&&OrchestrationStatus!=Running, to include all the other Terminal statuses aswell?

TsuyoshiUshio replied:
Done

adarsh1 commented on src/DurableTask.AzureStorage/AzureStorageOrchestrationService.cs:
```diff
@@ -833,17 +864,65 @@ async Task AbandonAndReleaseSessionAsync(OrchestrationSession session)
+            if (orchestrationState.OrchestrationStatus == OrchestrationStatus.Completed)
+            {
+                CorrelationTraceClient.TrackDepencencyTelemetry(dependencyActivity);
+            }
```
Why are these all separate Ifs? instead of a single if ((orchestrationState.OrchestrationStatus == OrchestrationStatus.Completed) ||
                (orchestrationState.OrchestrationStatus == OrchestrationStatus.Failed)) ?

TsuyoshiUshio replied:
Done

adarsh1 commented on src/DurableTask.AzureStorage/StringExtensions.cs:
```diff
@@ -0,0 +1,37 @@
+        /// <param name="s"></param>
+        public static string GetTargetClassName(this string s)
+        {
+            var index = s.IndexOf('+');
```
A custom object Creator or not interface based Activities may be used, in which case this logic may not hold

adarsh1 commented on src/DurableTask.AzureStorage/AzureStorageOrchestrationService.cs:
```diff
@@ -1036,9 +1115,17 @@ void ReleaseSession(string instanceId)
                 session.StartNewLogicalTraceScope();
+
+                // correlation 
+                string name = $"{TraceMessages.Activity} {((TaskScheduledEvent)session.MessageData.TaskMessage.Event)?.Name?.GetTargetClassName()}";
```
@simonporter to confirm here, but I don't believe this custom logic to get the class NAme would always hold, particularly in the case of a custom NameVersion object manager. Can this be moved one level up to The TaskOrchestrationManager in Core? In that case you would have access to the object manager and can get the class NAme more simply.

TsuyoshiUshio replied:
@adarsh1  Where can I find the TaskOrchestrationManager? I can't find it.

adarsh1 replied:
Sorry meant the TaskActivityDispatcher

adarsh1 commented on src/DurableTask.AzureStorage/Messaging/TaskHubQueue.cs:
```diff
@@ -105,7 +109,27 @@ async Task<MessageData> AddMessageAsync(TaskMessage taskMessage, OrchestrationIn
+                    currentTraceContext = currentTraceContext ?? current.CreateTraceContext();
+
+                    data.TraceContextStore = TraceContextStore.Create(currentTraceContext);
+                    rawContent = await messageManager.SerializeMessageDataAsync(data);
```
looks like data is the only thing changing between the if and else, should the raw content initialization simply stay out of this if else structure?

TsuyoshiUshio replied:
Done

adarsh1 commented on Test/DurableTask.AzureStorage.Tests/DurableTask.AzureStorage.Tests.csproj:
```diff
@@ -18,7 +18,6 @@
     <ProjectReference Include="..\..\src\DurableTask.Core\DurableTask.Core.csproj" />
     <ProjectReference Include="..\..\src\DurableTask.AzureStorage\DurableTask.AzureStorage.csproj" />
     <ProjectReference Include="..\DurableTask.Core.Tests\DurableTask.Core.Tests.csproj" />
-    <ProjectReference Include="..\DurableTask.Test.Orchestrations\DurableTask.Test.Orchestrations.csproj" />
```
Is this intentional?

TsuyoshiUshio replied:
Not there anymore.

adarsh1 commented on src/DurableTask.AzureStorage/DurableTask.AzureStorage.csproj:
```diff
@@ -35,4 +35,5 @@
   <ItemGroup>
     <ProjectReference Include="..\DurableTask.Core\DurableTask.Core.csproj" />
   </ItemGroup>
+
```
Undo if no change.

TsuyoshiUshio replied:
Done

TsuyoshiUshio commented:
Thank you @adarsh1 ! I'll do it as we discussed. I'll start this week. :) I appreciate your review.

TsuyoshiUshio commented on src/DurableTask.AzureStorage/AzureStorageOrchestrationService.cs:
```diff
@@ -612,6 +614,7 @@ static TaskHubInfo GetTaskHubInfo(string taskHub, int partitionCount)
                     session.StartNewLogicalTraceScope();
 
                     List<MessageData> outOfOrderMessages = null;
+                    var current = Activity.Current; // Correlation for checking the state of the Activity.Current. TODO remove
```
Removed

TsuyoshiUshio commented on src/DurableTask.Core/TraceContext.cs:
```diff
@@ -0,0 +1,70 @@
+//  ----------------------------------------------------------------------------------
+
+namespace DurableTask.Core
+{
```
I introduce the new structure.  I encapsulate activity to the TraceContext object. New name is 

* TraceContextBase (abstract class)
* W3CTraceContext  (implementation of the TraceContextBase)
* HttpCorrelationProtocolTraceContext (Implementation of the TraceContextBase)

For enabling this, I create a Factory to refer the configuration of the Protocol settings. 
These two Contexts have limited property. Some property is not serialized to reduce the size of the message. It requires special serialization, It is handled on TraceContextStore

TsuyoshiUshio commented on src/DurableTask.Core/TraceMessages.cs:
```diff
@@ -0,0 +1,28 @@
+{
+    /// <summary>
+    /// TraceMessages for Distributed Tracing
+    /// </summary>
```
Done

TsuyoshiUshio commented on src/DurableTask.Core/TraceContextStore.cs:
```diff
@@ -0,0 +1,63 @@
+    /// For the Newton.json 7.0.1 requires JsonSerializerSettings. It will all so requires DurableTask users.
+    /// To avoid this breaking change, I provide TraceContextStore which support the serialization/deserialization of the <see cref="TraceContext"/>
+    /// </summary>
+    public class TraceContextStore
```
I removed the TraceContextStore and make it available for TraceContextBase. I implement it as property and static method. I tried to use constructor, however, it requires additional logic to move restored object to the target object. It could be error prone. I create a Restore method simply pass the json string it creates TraceContextBase subclass object.

TsuyoshiUshio commented:
Hi @adarsh1 , 

I fix all your request. Review it again, please. We might need to check if there is no side effect for ServiceBus side. (Maybe we can use CI).

adarsh1 commented on src/DurableTask.Core/TraceContextBase.cs:
```diff
@@ -0,0 +1,160 @@
+        /// This property is not serialized.
+        /// </summary>
+        [JsonIgnore]
+        public Activity CurrentActivity { get; set; } // TODO make it protected after refactoring is done. 
```
make protected, or remove TODO

TsuyoshiUshio replied:
I double check the dependency. This property depend on the subclass, factory, and test class. I hope only these three can use this property. I madeit internal and removed TODO.

adarsh1 commented on src/DurableTask.Core/W3CTraceContext.cs:
```diff
@@ -0,0 +1,161 @@
+            Traceparent = CurrentActivity.GetTraceparent();
+            Tracestate = CurrentActivity.GetTracestate();
+            ParentSpanId = CurrentActivity.GetParentSpanId();
+            // ParentId = activity.Id, // TODO check if it is not necessary
```
Is this still necessary?

TsuyoshiUshio replied:
Removed

adarsh1 commented on src/DurableTask.Core/HttpCorrelationProtocolTraceContext.cs:
```diff
@@ -0,0 +1,76 @@
+        public override TimeSpan Duration => CurrentActivity?.Duration ?? DateTimeOffset.UtcNow - StartTime;
+
+        /// <inheritdoc />
+        public override string TelemetryId => throw new NotImplementedException(); // TODO Implement the HTTPCorrelation Protocol
```
Is this Safe? Would things fail if HttpCorrelationProtocol is used?

TsuyoshiUshio replied:
HttpCorrelationProtocol is not fully tested. App Insights team encourage to use W3C now. So I start implement/test with W3C. However, I implement HttpCorrelationProtocol. It should work, however, to make sure it works, it requires Scenario Test with TelemetryInitializer for HttpCorrelationProtocol.

adarsh1 commented on src/DurableTask.Core/TraceContextBase.cs:
```diff
@@ -0,0 +1,160 @@
+        public string SerializableTraceContext =>
+            JsonConvert.SerializeObject(this, new JsonSerializerSettings()
+            {
+                TypeNameHandling = TypeNameHandling.Objects,
```
We really should be using a custom JsonConverter here, if we need special handling. 
If we don't have the time to do that for now maybe a TODO would be ok?

TsuyoshiUshio replied:
I add TODO.

TsuyoshiUshio commented:
Thank you for your review, @adarsh1 , Hi, @cgillum , can I add my github account to the Azure DevOps task? I'd like to test this PR doesn't broke tests exists.

cgillum commented:
@TsuyoshiUshio yes, please go ahead.

TsuyoshiUshio commented:
@adarsh1 , @cgillum I copy and create a specific CI for this branch which anytime, we kick off the test. Some test fails. I'll fix it and push again.

TsuyoshiUshio commented:
Hi @adarsh1 and @cgillum , 
I solved the CI error. I introduce NullObject pattern. Some case, (e.g Rewind, Actor) it is difficult to keep the correlation. Also, For Actor implementation, I have not investigate the code yet. so I need to avoid to produce errors such unsupported scenario. Could you review it, please? Now All CI check has passed. https://durabletaskframework.visualstudio.com/Durable%20Task%20Framework%20CI/_build/results?buildId=191

TsuyoshiUshio commented:
FYI @cgillum will review next week. :)

cgillum commented on Correlation.Samples/ContinueAsNewScenario.cs:
```diff
@@ -0,0 +1,76 @@
+            {
+                await host.StartAsync();
+                var activity = new Activity("Start Orchestration");
+#pragma warning disable 618
```
Two things:

* Why are we disabling this warning here? If it is a valid reason, there should be a comment explaining this.
* You should restore this warning after disabling it so that you don't hide other legitimate warnings (e.g. `#pragma warning restore 618`)

Same comment for other places where warnings are being disabled.

TsuyoshiUshio replied:
I've done it except for the DurableTaskCorrelationTelemetryInitializer.cs . It is a special class. I'll mention it later. The warning suppression required if you want to use w3c trace context. I talked with Liudmila about it.  Currently it is required. However, in the near feature, it will be integrated. (maybe summer) then we can remove the warning and update the nuget package.

cgillum commented on Correlation.Samples/DurableTaskCorrelationTelemetryInitializer.cs:
```diff
@@ -0,0 +1,151 @@
+    using System;
+    using System.Collections.Generic;
+    using System.ComponentModel;
+    using System.Diagnostics;
```
nit: the rest of the codebase puts `System` namespaces at the top of the list. I suggest you do the same. You can update your Visual Studio settings to do this automatically.

TsuyoshiUshio replied:
Done

cgillum commented on Correlation.Samples/DurableTaskCorrelationTelemetryInitializer.cs:
```diff
@@ -0,0 +1,151 @@
+        {
+            OperationTelemetry operationTelemetry1 = telemetry as OperationTelemetry;
+            W3CTraceContext context = (W3CTraceContext)contextBase;
+            bool flag = operationTelemetry1 != null;
```
You should not name a boolean variable `flag` because that is not at all descriptive of its purpose. Please use a more meaningful variable name.

TsuyoshiUshio replied:
I'll explain later.

cgillum commented on Correlation.Samples/DurableTaskCorrelationTelemetryInitializer.cs:
```diff
@@ -0,0 +1,151 @@
+            W3CTraceContext context = (W3CTraceContext)contextBase;
+            bool flag = operationTelemetry1 != null;
+            if (flag)
+                flag = ((flag ? 1 : 0) & (!(operationTelemetry1 is DependencyTelemetry dependencyTelemetry) || !(dependencyTelemetry.Type == "SQL") ? 1 : (!dependencyTelemetry.Context.GetInternalContext().SdkVersion.StartsWith("rdddsc", StringComparison.Ordinal) ? 1 : 0))) != 0;
```
Please put curly braces around the body of `if` statements. That is the convention for the durabletask codebase and all new code should follow it.

cgillum replied:
Also, this line is too long. Please break it down into multiple lines so that it is easier to read.

cgillum replied:
Same feedback for the other code in this file.

TsuyoshiUshio replied:
I'll explain later.

cgillum commented on Correlation.Samples/DurableTaskCorrelationTelemetryInitializer.cs:
```diff
@@ -0,0 +1,151 @@
+                operationTelemetry1.Properties["w3c_tracestate"] = context.Tracestate;
+            }
+
+            var traceparent = context.Traceparent.ToTraceparent();
```
nit: `var` should only be used when the type of the argument is obvious from the right-hand expression. Please use the actual type name here instead of `var`.

TsuyoshiUshio replied:
Done

cgillum commented on Correlation.Samples/DurableTaskCorrelationTelemetryInitializer.cs:
```diff
@@ -0,0 +1,151 @@
+
+        internal static void UpdateTelemetry(ITelemetry telemetry, TraceContextBase contextBase)
+        {
+            OperationTelemetry operationTelemetry1 = telemetry as OperationTelemetry;
```
Why is there a `1` suffix in `operationTelemetry1`? This is generally not good coding practice. I'm guessing this code might have been autogenerated and/or decompiled from somewhere? Regardless, please remove the suffix or use a more descriptive variable name. Otherwise it can confuse the reader (for example, it makes me expect to see `operationTelemetry2` somewhere.

TsuyoshiUshio replied:
I'll explain it later

cgillum commented on Correlation.Samples/DurableTaskCorrelationTelemetryInitializer.cs:
```diff
@@ -0,0 +1,151 @@
+                        if (!(key == "w3c_parentSpanId"))
+                        {
+                            if (key == "w3c_tracestate" && telemetry is OperationTelemetry operationTelemetry2)
+                                operationTelemetry2.Properties["w3c_tracestate"] = tag.Value;
```
It looks like this code may have been copied from somewhere because it makes lots of assumptions about property names. If this code was copied from somewhere, you should add comments explaining where it came from so taht the reader can understand where these key names are coming from.

TsuyoshiUshio replied:
Yes. As you see, it is coming from other code. Unfortunately, the code is not extensible and the code is I can't completely understand the reason that is used for generic. I already have this comment on this class.  What I did was, The original telemetryInitializer hook the dependency telemetry and update the telemetry. Also, it expect the behavior Not like durable (no restore state of telemetry). For example, when the orchestrator execute twice, then there is no Activity instance. In this case this telemetry initializer create a new Activity that is not correlated. So I use CorrelateionTraceContext in this case. I try not to change the original one there is no test code and there is some behavior that I can't understand. 

However, if you want to make it clean code, I'll talk with Liudmila and make it "Clean code" :) 

```
    /// <summary>
    /// TelemetryInitializer with W3C TraceContextBase support.
    /// Customized code for <see cref="W3COperationCorrelationTelemetryInitializer"/>
    /// </summary>
```

cgillum commented on Correlation.Samples/MultiLayerOrchestrationWithRetryScenario.cs:
```diff
@@ -0,0 +1,99 @@
+            }
+        }
+    }
+    [KnownType(typeof(MultiLayeredOrchestrationChildWithRetry))]
```
nit: please add a newline here.

TsuyoshiUshio replied:
Done

cgillum commented on Correlation.Samples/StringExtensions.cs:
```diff
@@ -0,0 +1,45 @@
+
+    public static class StringExtensions
+    {
+        public static Traceparent ToTraceparent(this string traceparent)
```
nit: It seems like it should be `ToTraceParent` instead of `ToTraceparent` since `Trace` and `Parent` are distinct words.

TsuyoshiUshio replied:
This is debatable. I follow the naming rule of Acitivty W3C trace context. The class is coined it as 'Traceparent' also the spec is the same. (Although, VS and my ReSharper complain this.) :) However, if you prefer TraceParent, I'd happy to do it. 

 https://www.w3.org/TR/trace-context/

cgillum commented on Correlation.Samples/TerminationScenario.cs:
```diff
@@ -0,0 +1,72 @@
+
+        protected override async Task<string> ExecuteAsync(TaskContext context, string input) { 
+
+        // Wait for 5 min for terminate. 
```
nit: the whitespace on this method needs to be cleaned up - e.g. around the `{` and the `//`.

TsuyoshiUshio replied:
Done

cgillum commented on src/DurableTask.AzureStorage/AzureStorageOrchestrationService.cs:
```diff
@@ -656,13 +659,39 @@ static TaskHubInfo GetTaskHubInfo(string taskHub, int partitionCount)
+                        if (parentTraceContext.OrchestrationTraceContexts.Count != 0)
+                            requestTraceContextBase = parentTraceContext.OrchestrationTraceContexts.Peek();
+                        requestTraceContextBase.IsReplay = true;
+                    }
```
Can all of this new code be put into a method?

TsuyoshiUshio replied:
Let me clarify this. You mean the new code of DistributedTracing on the AzureStorageOrchestrationService should be extract to methods? Or just L682-684?

cgillum commented on src/DurableTask.AzureStorage/AzureStorageOrchestrationService.cs:
```diff
@@ -833,17 +862,67 @@ async Task AbandonAndReleaseSessionAsync(OrchestrationSession session)
+                {
+                    currentTraceContextBaseOnComplete = dependencyTraceContext.OrchestrationTraceContexts.Pop(); // Remove the current OrchestrationState(TraceContextBase) from the stack.
+
+                } else
```
nit: please put the `else` on its own line and remove the extra whitespace above this line.

TsuyoshiUshio replied:
Done

TsuyoshiUshio commented:
Hi @cgillum , I fixed some. Also I have something to discuss / clarify it. Could you see my comments for more details, please. However, In summary

* DurableTaskTelemetryInitializer is copied code with little fix. So I keep the original code. I already have a comment for the reference of original. Would you want to clean up the code? 
* Naming of Traceparent is coming from Activity W3C naming convention. Original spec is the same. https://www.w3.org/TR/trace-context/ However, if you want to change it I'd happy to do it. 
* Let me clarify, Do I need to extract methods that is all code related Distributed Tracing on AzureStorageOrchestrationService? Or just three lines that you specify? 

After getting answer from you, I'd happy to re-refactor it.

TsuyoshiUshio commented:
Hi @cgillum , 

I fixed three things that you mentioned. 

* DurableTaskTelemetryIntializer is taken from the original source and test it. 
* extract method that you indicate that. 
* Naming of Traceparent/Tracestate. I change it to TraceParent/TraceState

cgillum commented on src/DurableTask.AzureStorage/AzureStorageOrchestrationService.cs:
```diff
@@ -656,13 +659,26 @@ static TaskHubInfo GetTaskHubInfo(string taskHub, int partitionCount)
+                    var isReplaying = session.RuntimeState.ExecutionStartedEvent?.IsPlayed ?? false;
+
+                    var firstMessage = session.CurrentMessageBatch.FirstOrDefault();
+                    TraceContextBase parentTraceContext = null;
```
A message batch may contain messages from multiple sources. For example, you could have both an external event message and a task completed message in the same batch. Both will have different contexts. Depending on which message is received first in the batch, you will restore just one context and the other context will be lost. Is this a problem? If so, how do we fix this problem?

TsuyoshiUshio replied:
I did some experiment for the case. I reproduce two messages issue. I change the logic not to break the orchestration. However, I found that sometimes, external event only message could come. In this case, we need to ignore it, in terms of telemetry.

TsuyoshiUshio replied:
I do refactor this as well according to your advice below.

TsuyoshiUshio replied:
I also refactor it to make it clear the intention.

cgillum commented on src/DurableTask.AzureStorage/AzureStorageOrchestrationService.cs:
```diff
@@ -723,10 +739,30 @@ static TaskHubInfo GetTaskHubInfo(string taskHub, int partitionCount)
+                // In case of termination, it will cause an error. There is no OrchestrationTraceContexts. 
+                if (parentTraceContext.OrchestrationTraceContexts.Count != 0)
+                    requestTraceContextBase = parentTraceContext.OrchestrationTraceContexts.Peek();
+                requestTraceContextBase.IsReplay = true;
```
I suggest you refactor this so that the high-level meaning is more clear.

```
bool noCorrelation = parentTraceContext.OrchestrationTraceContexts.Count == 0;
if (noCorrelation)
{
 // Terminate, external events, etc. are examples of messages that might not contain any trace context.
 // In those cases, we just return an empty trace context and continue on.
 return TraceContextFactory.Empty;
}

requestTraceContext = parentTraceContext.OrchestrationTraceContexts.Peek();
requestTraceContext.IsReplay = true;
return requestTraceContextBase;
```

TsuyoshiUshio replied:
Done. Thank you for good advice.

TsuyoshiUshio commented on src/DurableTask.AzureStorage/AzureStorageOrchestrationService.cs:
```diff
@@ -656,13 +659,26 @@ static TaskHubInfo GetTaskHubInfo(string taskHub, int partitionCount)
 
+                    var isReplaying = session.RuntimeState.ExecutionStartedEvent?.IsPlayed ?? false;
+
+                    var correlatedMessage = GetCorrelatedMessage(session.CurrentMessageBatch);
```
This var should be MessageData

TsuyoshiUshio commented on src/DurableTask.AzureStorage/AzureStorageOrchestrationService.cs:
```diff
@@ -656,13 +659,26 @@ static TaskHubInfo GetTaskHubInfo(string taskHub, int partitionCount)
+                        parentTraceContext = TraceContextBase.Restore(correlatedMessage?.SerializableTraceContext);
+                    }
+
+                    TraceContextBase requestTraceContextBase = TraceContextFactory.Default;
```
Make it Empty instead of Default

TsuyoshiUshio commented on src/DurableTask.AzureStorage/AzureStorageOrchestrationService.cs:
```diff
@@ -656,13 +659,26 @@ static TaskHubInfo GetTaskHubInfo(string taskHub, int partitionCount)
+                    TraceContextBase parentTraceContext = TraceContextFactory.Default;
+                    if (correlatedMessage != null)
+                    {
+                        parentTraceContext = TraceContextBase.Restore(correlatedMessage?.SerializableTraceContext);
```
? is not needed.

TsuyoshiUshio commented on src/DurableTask.AzureStorage/AzureStorageOrchestrationService.cs:
```diff
@@ -723,10 +739,50 @@ static TaskHubInfo GetTaskHubInfo(string taskHub, int partitionCount)
+            return null;
+        }
+
+        static TraceContextBase UpdateRequestTraceContextBase(bool isReplaying, MessageData firstMessage, TraceContextBase requestTraceContextBase, TraceContextBase parentTraceContext)
```
change name to correlatedMessage instead of firstMessage

TsuyoshiUshio commented:
@cgillum  As we talked, I re-target it, and remove the conflict.

cgillum commented on samples/Correlation.Samples/Correlation.Samples.csproj:
```diff
@@ -0,0 +1,22 @@
+
+  <PropertyGroup>
+    <OutputType>Exe</OutputType>
+    <TargetFramework>net451</TargetFramework>
```
Some comment about targeting .NET Core.

TsuyoshiUshio replied:
change it to netcoreapp2.1

TsuyoshiUshio replied:
Done

cgillum commented on samples/Correlation.Samples/Program.cs:
```diff
@@ -0,0 +1,41 @@
+            CorrelationSettings.Current.EnableDistributedTracing = true;
+            // InvokeScenario(typeof(HelloOrchestrator), "50", 50); // HelloWorldScenario.cs;
+            // InvokeScenario(typeof(SubOrchestratorOrchestration), "SubOrchestrationWorld", 50);  // SubOrchestratorScenario.cs;
+            // InvokeScenario(typeof(RetryOrchestration), "Retry Scenario", 50); // RetryScenario.cs;
```
Are these supposed to be commented out?

TsuyoshiUshio replied:
@cgillum Yes. It is a sample of other scenarios. People can comment out/in, then can try all scenario.

cgillum commented on samples/Correlation.Samples/TelemetryActivator.cs:
```diff
@@ -0,0 +1,87 @@
+
+    public class TelemetryActivator
+    {        
+        static TelemetryClient telemetryClient;
```
Same comment about whether this should be static.

TsuyoshiUshio replied:
Done

cgillum commented on src/DurableTask.AzureStorage/AzureStorageOrchestrationService.cs:
```diff
@@ -1149,23 +1369,34 @@ public async Task CompleteTaskActivityWorkItemAsync(TaskActivityWorkItem workIte
             if (!this.activeActivitySessions.TryGetValue(workItem.Id, out session))
             {
-                // The context does not exist - possibly because it was already removed.
+                // The contextBase does not exist - possibly because it was already removed.
```
Is this rename a mistake? This code is not related to tracing. Can you please revert this everywhere?

TsuyoshiUshio replied:
Done

cgillum replied:
There are many more places in this code that still need to be fixed. Please fix them all.

TsuyoshiUshio replied:
Done

cgillum commented on src/DurableTask.AzureStorage/Messaging/TaskHubQueue.cs:
```diff
@@ -11,16 +11,21 @@
 //  limitations under the License.
 //  ----------------------------------------------------------------------------------
 
+#pragma warning disable 618
```
Why is this warning being disabled? Can you please add a comment?

TsuyoshiUshio replied:
Done

cgillum commented on src/DurableTask.AzureStorage/Messaging/TaskHubQueue.cs:
```diff
@@ -106,7 +111,12 @@ async Task<MessageData> AddMessageAsync(TaskMessage taskMessage, OrchestrationIn
+                CorrelationTraceClient.Propagate(
+                    () => { data.SerializableTraceContext = GetSerializableTraceContext(taskMessage); });
+                
+                string rawContent = await messageManager.SerializeMessageDataAsync(data);
```
Why did you remove the `this.` prefix? Can you please revert this change?

TsuyoshiUshio replied:
Done

cgillum replied:
This is not resolved in the latest code.

TsuyoshiUshio replied:
Sorry, I fixed.

cgillum commented on src/DurableTask.Core/CorrelatedException.cs:
```diff
@@ -0,0 +1,53 @@
+    /// <summary>
+    /// A class that includes an exception and correlation information for sending telemetry
+    /// </summary>
+    public class CorrelatedException
```
Let's come up with a different name for this class, like `CorrelatedExceptionDetails`. Otherwise people will accidentally think this is an exception that can be thrown.

TsuyoshiUshio replied:
Done

cgillum commented on src/DurableTask.Core/DurableTask.Core.csproj:
```diff
@@ -17,4 +17,14 @@
+  </ItemGroup>
+
+  <ItemGroup>
+    <PackageReference Include="Microsoft.ApplicationInsights.DependencyCollector" Version="2.8.0" />
```
Can you add a TODO comment explaining that we want to remove this dependency in a future PR?

TsuyoshiUshio replied:
Done

cgillum commented on src/DurableTask.Core/TaskOrchestrationContext.cs:
```diff
@@ -451,6 +451,8 @@ public void FailOrchestration(Exception failure)
             string reason = failure.Message;
             string details;
+            // correlation 
+            CorrelationTraceClient.TrackException(failure);
```
Do you need `CorrelationTraceClient.Propagate(() => ...)` here?

TsuyoshiUshio replied:
Done

cgillum commented on src/DurableTask.Core/TraceContextBase.cs:
```diff
@@ -0,0 +1,198 @@
+                var restored = JsonConvert.DeserializeObject(
+                    json,
+                    traceContextType,
+                    new JsonSerializerSettings()
```
Can you add a TODO to cache these serializer settings, or get them from somewhere else?

TsuyoshiUshio replied:
Done

TsuyoshiUshio commented on Test/DurableTask.AzureStorage.Tests/Correlation/LIstExtensionsTest.cs:
```diff
@@ -0,0 +1,104 @@
+    using Microsoft.VisualStudio.TestTools.UnitTesting;
+
+    [TestClass]
+    public class ListExtensionsTest
```
Thank you for finding this!

TsuyoshiUshio replied:
Solved

TsuyoshiUshio commented on Test/DurableTask.AzureStorage.Tests/Correlation/LIstExtensionsTest.cs:
```diff
@@ -0,0 +1,104 @@
+﻿using System;
```
Sorry about that. I added it for all code that doesn't have it.

TsuyoshiUshio commented on Test/DurableTask.AzureStorage.Tests/Correlation/LIstExtensionsTest.cs:
```diff
@@ -0,0 +1,104 @@
+{
+    using Microsoft.ApplicationInsights.DataContracts;
+    using Microsoft.ApplicationInsights.Extensibility.Implementation;
+    using Microsoft.VisualStudio.TestTools.UnitTesting;
```
Sorry again, I forget to change it. I fixed all of others as well.

TsuyoshiUshio commented on Test/DurableTask.AzureStorage.Tests/Correlation/LIstExtensionsTest.cs:
```diff
@@ -0,0 +1,104 @@
+            return telemetry;
+        }
+    
+}
```
I've done it.

TsuyoshiUshio commented on Test/DurableTask.AzureStorage.Tests/Correlation/StubTelemetryChannel.cs:
```diff
@@ -0,0 +1,91 @@
+
+    /// <summary>
+    /// A stub of <see cref="ITelemetryChannel"/>.
+    /// This is the copy of the https://github.com/Microsoft/ApplicationInsights-dotnet/Test/TestFramework/Shared/StubTelemetryClient
```
Thanks. Done.

TsuyoshiUshio commented on Test/DurableTask.AzureStorage.Tests/Correlation/StubTelemetryChannel.cs:
```diff
@@ -0,0 +1,91 @@
+    /// <summary>
+    /// A stub of <see cref="ITelemetryChannel"/>.
+    /// This is the copy of the https://github.com/Microsoft/ApplicationInsights-dotnet/Test/TestFramework/Shared/StubTelemetryClient
+    public sealed class StubTelemetryChannel : ITelemetryChannel
```
It means No Operation? Anyway done.

TsuyoshiUshio commented on Test/DurableTask.AzureStorage.Tests/Correlation/TelemetryActivator.cs:
```diff
@@ -0,0 +1,93 @@
+
+    public class TelemetryActivator
+    {        
+        static TelemetryClient telemetryClient;
```
@cgillum  Originally, I intended to be the telemetryClient should be static. There is no problem to switch to instance one since this is test class. However,  when I remove static and make it private, it start to fail only W3C. I debug some however, I can't get root cause until now. I'll keep on investigating it.

TsuyoshiUshio replied:
I investigate the issue, I found two thing.  When I change to instance variables, I notice that.
* It fails on the second DataRow of DataTestMethod (not related W3C) 

In this case, the second test execution has failed since there is no telemetry. I haven't get the root cause until now, however, I added a comment. 
```
        [DataTestMethod]
        [DataRow(Protocol.HttpCorrelationProtocol)]
        [DataRow(Protocol.W3CTraceContext)]
        public async Task MultipleParentScenarioAsync(Protocol protocol)
```

cgillum replied:
Your comment says "TODO: Should be static". Did you mean to say "TODO: Should _not_ be static"?

The test failure sounds like a bug because static state is not cleaned up in between tests. In other words, the first test is initializing state that your second test is depending on. All tests should be decoupled from each other. I'm okay if we skip investigating for now as long as we remember to follow-up.

_(side note: in your reply you said "I haven't get the root cause **until now**". Did you mean to say "I haven't found the root cause **yet**"? I think you meant to say "**yet**" here. The meaning of **until now** is that you did find the root cause just now. :))_

TsuyoshiUshio replied:
Thank you for the comment. I appreciate it for the grammar correction. :) 
BTW, I finally figure out the root cause of the issue.  Please see this commit. 
https://github.com/TsuyoshiUshio/durabletask/commit/adf9587ecbdbcf90dd2e9cb8240820c00f11f22e

The problem was, From the client side, it subscribe trackRequest/Dependency/ExceptionTelemetryAction on CorrelationTraceClient.SetUp() method. it was null checked if listernerSubscription. So if you execute SetUp() twice with instance member of TelemetryClient, the second execution of SetUp() is ignored. However, TelemetryClient instance is already gone since it is instance memeber and object has gone. 

The next question is, SetUp() method should not be executed twice. How can I avoid the users to execute SetUp() method twice? It works, however, if you execute SetUp several times, it increases subscribers.

TsuyoshiUshio replied:
Anyway, I already change it to non-static.

TsuyoshiUshio commented on Test/DurableTask.Core.Tests/TraceContextBaseTest.cs:
```diff
@@ -0,0 +1,168 @@
+            var dependencyContext = new Foo() { Comment = comment };
+            dependencyContext.TelemetryType = TelemetryType.Dependency;
+            return dependencyContext;
+        }
```
Done

TsuyoshiUshio commented on Test/DurableTask.Core.Tests/W3CTraceContextTest.cs:
```diff
@@ -0,0 +1,42 @@
+//  See the License for the specific language governing permissions and
+//  limitations under the License.
+//  ----------------------------------------------------------------------------------
+#pragma warning disable CS0618 
```
Done

TsuyoshiUshio commented on samples/Correlation.Sample.Tests/Correlation.Sample.Tests.csproj:
```diff
@@ -0,0 +1,17 @@
+﻿<Project Sdk="Microsoft.NET.Sdk">
+
+  <PropertyGroup>
+    <TargetFramework>net451</TargetFramework>
```
Done

TsuyoshiUshio replied:
Done

TsuyoshiUshio commented:
Hi @cgillum , 

I fix the issues that you requested. However, some need to discuss with you. I leave these as  unresovled comments. Could you have a look these three? 

1. TelmetryActiviator's TelemetryClient should be static, left a comment. 
2. samples/Correlation.Samples/Program.cs I leave the comments for represent the available demo scenario.
3.  How? (following picture) 
![image](https://user-images.githubusercontent.com/1390976/72671787-74147780-3a04-11ea-99b3-de2fda7b3d95.png)

cgillum commented on src/DurableTask.Core/StackExtensions.cs:
```diff
@@ -0,0 +1,25 @@
+﻿using System;
```
This new file needs the copyright header. Also, the `using` statements need to go inside the `namespace` section.

TsuyoshiUshio replied:
Done

cgillum commented:
Hi @TsuyoshiUshio - I have reviewed your latest updates and responded to your questions. There are also a few requested fixes that have not been made yet. Please take a look.

TsuyoshiUshio commented:
Hi @cgillum , I fixed them.  Only one concerns left. 

1. I fixed static TelemetryClient issue and find out the root cause.  I'd like to listen the opinion about the design decision. Please refer the comment. (I leave it  not resolved)

TsuyoshiUshio commented:
Hi @cgillum this is friendly reminder. :)

jkerak commented:
Hi. I'm not sure if this is the correct place to ask this question (I can't find a link to some kind of issue or feature request). Is there any kind of limit to number of executions that will be logged to AI? In some cases, I may have tens of thousands of activity function executions within a single orchestrator execution.

cgillum commented on src/DurableTask.Core/CorrelationTraceClient.cs:
```diff
@@ -0,0 +1,129 @@
+    /// Delegate sending telemetry to the other side.
+    /// Mainly send telemetry to the Durable Functions TelemetryClient
+    /// </summary>
+    public class CorrelationTraceClient
```
nit: this should be a `static` class since it has no instance members.

TsuyoshiUshio replied:
I've done that!

TsuyoshiUshio commented:
Hi @cgillum I fixed all of the issues.

cgillum commented:
Thanks! I will merge this now. Glad to get this first step finally checked in. :)

TsuyoshiUshio commented:
Hi @jkerak 

You are using tens of thousands of activity functions in a single orchestrator right? I'm not sure about the limitation of the durable side, however, Application Insights has some quota https://docs.microsoft.com/en-us/azure/azure-monitor/app/api-custom-events-metrics#limits This issue might be a good place to talk about correlation. https://github.com/Azure/azure-functions-durable-extension/issues/939

TsuyoshiUshio commented:
Thank you @cgillum  for kind support of this PR. I'll go in implementing durable functions side.
